ast/compile_test.go

```diff
@@ -2335,6 +2341,321 @@ elsekw {
+}`,
+		},
+		{
+			note: "rego.metadata argument in function call",
+			module: `package test
+
+p {
+	q(rego.metadata.chain())
+}
+
+q(s) {
+	s == ["test", "p"]
+}`,
+			exp: `package test
+
+p = true { 
+	__local2__ = [{"path": ["test", "p"]}]
+	__local1__ = __local2__
+	data.test.q(__local1__) 
+}
+
+q(__local0__) = true { 
+	equal(__local0__, ["test", "p"]) 
+}`,
+		},
+		{
+			note: "rego.metadata used in array comprehension",
+			module: `package test
+
+p = [x | x := rego.metadata.chain()]`,
+			exp: `package test
+
+p = [__local0__ | __local1__ = __local2__; __local0__ = __local1__] { 
+	__local2__ = [{"path": ["test", "p"]}]
+	true 
+}`,
+		},
+		{
+			note: "rego.metadata used in nested array comprehension",
+			module: `package test
+
+p {
+	y := [x | x := rego.metadata.chain()]
+	y[0].path == ["test", "p"]
+}`,
+			exp: `package test
+
+p = true { 
+	__local3__ = [{"path": ["test", "p"]}]; 
+	__local1__ = [__local0__ | __local2__ = __local3__; __local0__ = __local2__]; 
+	equal(__local1__[0].path, ["test", "p"]) 
+}`,
+		},
+		{
+			note: "rego.metadata used in set comprehension",
+			module: `package test
+
+p = {x | x := rego.metadata.chain()}`,
+			exp: `package test
+
+p = {__local0__ | __local1__ = __local2__; __local0__ = __local1__} { 
+	__local2__ = [{"path": ["test", "p"]}]
+	true 
+}`,
+		},
+		{
+			note: "rego.metadata used in nested set comprehension",
+			module: `package test
+
+p {
+	y := {x | x := rego.metadata.chain()}
+	y[0].path == ["test", "p"]
+}`,
+			exp: `package test
+
+p = true { 
+	__local3__ = [{"path": ["test", "p"]}]
+	__local1__ = {__local0__ | __local2__ = __local3__; __local0__ = __local2__}
+	equal(__local1__[0].path, ["test", "p"]) 
+}`,
+		},
+		{
+			note: "rego.metadata used in object comprehension",
+			module: `package test
+
+p = {i: x | x := rego.metadata.chain()[i]}`,
+			exp: `package test
+
+p = {i: __local0__ | __local1__ = __local2__; __local0__ = __local1__[i]} { 
+	__local2__ = [{"path": ["test", "p"]}]
+	true 
+}`,
+		},
+		{
+			note: "rego.metadata used in nested object comprehension",
+			module: `package test
+
+p {
+	y := {i: x | x := rego.metadata.chain()[i]}
+	y[0].path == ["test", "p"]
+}`,
+			exp: `package test
+
+p = true { 
+	__local3__ = [{"path": ["test", "p"]}]
+	__local1__ = {i: __local0__ | __local2__ = __local3__; __local0__ = __local2__[i]}
+	equal(__local1__[0].path, ["test", "p"]) 
+}`,
+		},
+	}
+
+	for _, tc := range tests {
+		t.Run(tc.note, func(t *testing.T) {
+			c := NewCompiler()
+			c.Modules = map[string]*Module{
+				"test.rego": MustParseModule(tc.module),
+			}
+			compileStages(c, c.rewriteRegoMetadataCalls)
+			assertNotFailed(t, c)
+
+			result := c.Modules["test.rego"]
+			exp := MustParseModuleWithOpts(tc.exp, ParserOptions{ProcessAnnotation: true})
+
+			if result.Compare(exp) != 0 {
+				t.Fatalf("\nExpected:\n\n%v\n\nGot:\n\n%v", exp, result)
+			}
+		})
+	}
+}
+
+func TestCompilerRewriteSelfCallsErrors(t *testing.T) {
```

Comment:
[nit] It's not "Self" anymore, is it?

🔍 these are the errors we do the rewrittenVar hack for? 🤔

Reply:
This isn't tied to `rego.metadata.*()`, I believe:

```console
$ opa eval -fpretty 'opa.runtime()[0].foo'
1 error occurred: 1:1: rego_type_error: undefined ref: __localq0__[0].foo
        __localq0__[0].foo
                    ^
                    have: 0
                    want (type): string
```
...maybe we keep the ugly error here and create an issue on improving errors of this sort?

Reply:
Oh, I didn't realize. Yea, then I'll back out of this and create a ticket for a general solution.

Reply:
Thanks!

Reply:
https://github.com/open-policy-agent/opa/issues/4577